src/v/cluster/partition_balancer_planner.cc

```diff
@@ -0,0 +1,385 @@
+        std::sort(ntp_on_node_sizes.begin(), ntp_on_node_sizes.end());
+        size_t index = 0;
+        _node_released_disk_size.insert({node_disk_space.node_id, 0});
+        while (double(
+                 _node_released_disk_size.at(node_disk_space.node_id)
+                 + node_disk_space.free_space)
+                   / double(node_disk_space.total_space)
+                 < double(1.0) - _config.max_disk_usage()
+               && index < ntp_on_node_sizes.size()) {
+            const auto& partition_to_move = ntp_on_node_sizes[index].second;
+            if (_moving_partitions.contains(model::ntp(
+                  partition_to_move.ns,
+                  partition_to_move.tp.topic,
+                  partition_to_move.tp.partition))) {
+                index++;
+                continue;
+            }
+            if (ntp_on_node_sizes[index].first == size_t(-1)) {
+                index++;
+                continue;
+            }
+
+            const auto& topic_metadata = _topic_table.local().topics_map().at(
+              model::topic_namespace(
+                partition_to_move.ns, partition_to_move.tp.topic));
+            if (!is_partition_movement_possible(
+                  topic_metadata.get_assignments()
+                    .find(partition_to_move.tp.partition)
+                    ->replicas)) {
+                index++;
+                continue;
+            }
+            auto stable_replicas = get_stable_replicas(
+              topic_metadata.get_assignments()
+                .find(partition_to_move.tp.partition)
+                ->replicas);
+            auto new_allocation_units = get_reallocation(
+              *topic_metadata.get_assignments().find(
+                partition_to_move.tp.partition),
+              topic_metadata,
+              stable_replicas,
+              topic_metadata.get_assignments()
+                .find(partition_to_move.tp.partition)
+                ->replicas,
+              ntp_on_node_sizes[index].first);
+            if (!new_allocation_units) {
+                vlog(
+                  clusterlog.info,
+                  "Can't reallocate {} {} {}",
+                  partition_to_move.ns,
+                  partition_to_move.tp.topic,
+                  partition_to_move.tp.partition);
+            } else {
+                auto ntp = model::ntp(
+                  partition_to_move.ns,
+                  partition_to_move.tp.topic,
+                  partition_to_move.tp.partition);
+                _moving_partitions.insert(ntp);
+                reassigments.emplace_back(ntp_reassignments{
+                  .ntp = ntp,
+                  .allocation_units = std::move(new_allocation_units.value())});
+                _cur_batch_size += ntp_on_node_sizes[index].first;
+            }
+            index++;
+            if (_cur_batch_size >= _config.batch_size()) {
+                return;
+            }
+        }
+    }
+}
+
+std::vector<ntp_reassignments>
```

Comment:
since this is a synchronous method would it cause reactor stalls for high partition counts ?

Reply:
I think that's guaranteed. But I think we can futurize it later (I assume it would be easy, correct me if I'm wrong) after profiling - right now it is not obvious to me where we should insert yield points.